### PR TITLE
fix promise catch chaining 

### LINKED ISSUE #45: Context not preserved after Promise.catch
```
const cls = require('continuation-local-storage')
const context = cls.createNamespace('context')
context.run(() => { 
  context.set('foo', 3); 
  Promise.resolve('test')
    .then(() => console.log(cls.getNamespace('context').get('foo'))) 
})

=> 3

const cls = require('continuation-local-storage')
const context = cls.createNamespace('context')
context.run(() => { 
  context.set('foo', 3); 
  Promise.resolve('test')
    .catch(() => {})
    .then(() => console.log(cls.getNamespace('context').get('foo'))) 
})

=> undefined
````

### PR DESCRIPTION
**Pull request by hayes**
fixes #45 and #46 

this fixes catch chaining when there was no error (when it behaves as a proxy)
this also ensures that after hooks gets run even when handler throws an error.

### CODE REVIEW

**Comment by Qard** [comment]
:clap: LGTM!

**Comment by hayes** [comment]
test is broken, need to fix it real quick

**Comment by hayes** [comment]
there we go, test is fixed

File: index.js [hunk 1]
```diff
@@ -391,6 +393,12 @@ function wrapPromise() {
     return function wrappedThen() {
       var promise = this;
       var next = original.apply(promise, Array.prototype.map.call(arguments, bind));
+
+      next.__asl_wrapper = function proxyWrapper(ctx, fn, val, last) {
+        promise.__asl_wrapper(ctx, function () {}, null, next)
+        return next.__asl_wrapper(ctx, fn, val, last)
```
Comment (hayes) [inline index.js]: @Qard how do you feel about this? feels really dirty to have next.__asl_wrapper call a function that will replace the __asl_wrapper function on next, and then invoke that replaced function.  I couldn't think of a clean way around this, but open to suggestions

Comment (Qard) [inline index.js]: If it works, it works, I guess. :see_no_evil: :hear_no_evil: :speak_no_evil: 

Unless you have a better idea, I think we can go with this.

**Comment by hayes** [comment]
@Quard do you have publish permissions for this on npm?

**Comment by taion** [comment]
Oops. I'm embarrassed I forgot about `.catch()` the first time around.

**Comment by Qard** [comment]
I think so, want me to publish it? I'll be back at my computer soon.

**Pull request merged**